### PR TITLE
Define and Use the New AdapterFault struct for envoy adapter

### PR DESCRIPTION
**Pull request by kansalarchit**
Define a new AdapterFault struct for envoy adapter. 
Replace the bool fault with the new AdapterFault to make fault reporting dynamic.

### CODE REVIEW

**Comment by theganyo** [comment]
The copyright should be 2021. But other than that, this struct isn't being used anywhere and has a TODO, so it seems incomplete. What's the context on this?

**Comment by theganyo** [comment]
It's hard to tell if this is PR ready for review or not. Can we move it to "draft" if not, and then back when it's ready?

**Comment by kansalarchit** [comment]
The PR is ready for review. Please take a look.

**Review by theganyo — CHANGES_REQUESTED** [review]

File: server/header_context.go
Comment (kansalarchit) [reply server/header_context.go:109]: Sure, Done.

File: server/header_context.go
Comment (kansalarchit) [reply server/header_context.go:117]: Sure, added a conditional local assignment.

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:359]: Documented that the values can be optional, also added a method that takes just the rpcCode to construct AdapterFault.

File: server/header_context.go
Comment (kansalarchit) [reply server/header_context.go:117]: sure, Done.

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:359]: Sure, documented as well as added a method for CreateAdapterFaultWithRpcCode.

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go]: 1) I agree, this reduces the readability. This was documented in go/eg4a-runtime-fault-detection and therefore I went ahead and implemented it. Adding call-site param documentation to improve readability.

2) We don't have any use-case, changing the arg to AdapterFault everywhere.

Thanks for the suggestions!

File: config/env_spec_request.go
Comment (emschwar) [inline config/env_spec_request.go:357]: My intention was for this (and similar return values) to be returned as `error`. Although we do not intend to use the `Error()` function from the `error` interface directly, implementing it would nonetheless allow more conventional Go here, I think.

Comment (theganyo) [reply config/env_spec_request.go:357]: +1 to using as `error`

File: server/authorization.go
Comment (emschwar) [inline server/authorization.go:609]: The `adapterFault` might be better off passed in here as an `error`, and you can test casting it to `fault.AdapterFault` here, or in the `createConditionalEnvoyDenied`. WDYT?

File: server/authorization.go
Comment (emschwar) [inline server/authorization.go:77]: After reading this, I tend to agree that the prior versions were more readable. I don't think it's required, but if you wanted to bring them back, I don't think that was an essential part of the original design.

File: config/env_spec_request_test.go
Comment (rockspore) [reply config/env_spec_request_test.go:368]: +1

Or it should return both a bool and error?

Comment (theganyo) [reply config/env_spec_request_test.go:368]: Actually, now I'm not sure why we changed this function. We only call it from one place (I believe)... could not we have left it as a boolean and just generate the error at the caller if the result is false?

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:357]: I had used error initially but changed it later on due to misunderstanding an earlier comment. Changing it back to `error`. Thanks!

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:451]: Done.

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:489]: Done.

File: config/env_spec_request_test.go
Comment (kansalarchit) [reply config/env_spec_request_test.go:368]: Renamed it to Authenticate() and replaced the usages as suggested by Eric. 
 I think Authenticate() returning an error is more axiomatic way of doing things instead of caller doing the work simply because there is only one caller `so far`.
Thanks!

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go:609]: I had used error initially but changed it later on due to misunderstanding an earlier comment. Changing it back to error. Thanks!

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go:77]: I think with the call-site documentation in place, having a single handleFault looks good too.

**Review by theganyo — APPROVED** [review]
Please fix linter error and rebase with master. Other than that, LGMT.

File: config/env_spec_request.go
Comment (rockspore) [inline config/env_spec_request.go:483]: This method name also feels very like it should return a boolean, can we rename it to maybe `verifyAuthenticationRequirements`?

File: fault/fault.go
Comment (rockspore) [inline fault/fault.go:46]: `()` in the comment is not needed. Same for other places.

File: fault/fault_test.go
Comment (rockspore) [inline fault/fault_test.go:31]: The descriptions are better to be something more descriptive about what you are testing in this case :-)

File: fault/fault_test.go
Comment (rockspore) [inline fault/fault_test.go:48]: This is not needed. And if you are iterating over tests,  the standard pattern in the loop should be:
```
t.Run(test.Desc, func(t *testing.T) {
    // your existing testing logic
})
```

File: fault/fault_test.go
Comment (rockspore) [inline fault/fault_test.go:65]: Same here.

File: fault/fault_test.go
Comment (rockspore) [inline fault/fault_test.go:81]: Same here.

File: server/authorization.go
Comment (rockspore) [inline server/authorization.go:262]: line breaking for arguments is not recommended in go. Please convert it to the one-liner even if it's very long.

Comment (rockspore) [reply server/authorization.go:262]: Let me know if I am wrong though ;-)

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:483]: sure, Done.

File: fault/fault.go
Comment (kansalarchit) [reply fault/fault.go:46]: Thanks for informing, done.

File: fault/fault_test.go
Comment (kansalarchit) [reply fault/fault_test.go:31]: sure, updated.

File: fault/fault_test.go
Comment (kansalarchit) [reply fault/fault_test.go:48]: Thanks for pointing out, Done.

File: fault/fault_test.go
Comment (kansalarchit) [reply fault/fault_test.go:65]: Done.

File: fault/fault_test.go
Comment (kansalarchit) [reply fault/fault_test.go:81]: Done.

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go:262]: Thanks for pointing out. 
From what I can find (refer: https://stackoverflow.com/a/34848928) my use of line-breaks works just fine and it makes code much more readable.

**Comment by rockspore** [comment]
The linting failed so please fix it. Otherwise LGTM. Not approving it right now because once you update the PR the old approval will become ineffective.

**Comment by kansalarchit** [comment]
Fixed linter and resolved all comments, Please take a look. Thanks!

File: server/authorization.go
Comment (rockspore) [inline server/authorization.go:610]: Why err.Error()? err is sufficient for %v

Comment (kansalarchit) [reply server/authorization.go:610]: I assumed that printing the error string might be better. I can change it if you prefer, let me know.

Comment (rockspore) [reply server/authorization.go:610]: IIUC, they are the same. So err.Error() is redundant.

Comment (kansalarchit) [reply server/authorization.go]: I see, done.

**Review by rockspore — APPROVED** [review]

**Review by theganyo — DISMISSED** [review]

File: config/env_spec_request_test.go
Comment (emschwar) [inline config/env_spec_request_test.go:49]: nit: this would be a good time to fix this typo (should read: `verifyAuthenticationRequirements`) :joy:

File: server/authorization.go
Comment (emschwar) [reply server/authorization.go:77]: If the others are fine with is, I am too. Thanks for taking a look!

File: server/authorization.go
Comment (emschwar) [inline server/authorization.go:167]: nit: idiomatic Go would use `err` here.

**Review by emschwar — DISMISSED** [review]
A few minor nitpicks, but overall LGTM

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go]: Done.

File: config/env_spec_request_test.go
Comment (kansalarchit) [reply config/env_spec_request_test.go:49]: haha..thanks for pointing out, Done!

**Review by rockspore — DISMISSED** [review]

File: config/env_spec_request.go
Comment (emschwar) [inline config/env_spec_request.go:355]: nit: returning nil on success is assumed with Go; maybe rewrite this (and others) as "return error if not found, or not verified"

File: config/env_spec_request.go
Comment (emschwar) [inline config/env_spec_request.go:450]: Ditto here-- also,  you have two lines that basically say the same thing. Pick one? :)

File: config/env_spec_request.go
Comment (emschwar) [inline config/env_spec_request.go:488]: Ditto here.

File: fault/fault.go
Comment (emschwar) [inline fault/fault.go:47]: Nit: Conventional Go would use `New` vs `Create` here, and below.

**Review by emschwar — APPROVED** [review]
A few minor nitpicks here and there, but basically, it looks good to me!

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:355]: Done

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:450]: Rephrased.
The two statements are different. First says, nil if met and second says empty/disabled is also considered met.

File: config/env_spec_request.go
Comment (kansalarchit) [reply config/env_spec_request.go:488]: removed the comment, private function.

File: fault/fault.go
Comment (kansalarchit) [reply fault/fault.go:47]: Thanks for the info, done.

**Comment by kansalarchit** [comment]
Approvals expired, please take a look again.

File: server/authorization.go
Comment (emschwar) [reply server/authorization.go:262]: @rockspore is correct here; prefer to not break lines unless it is absolutely required. Our style prefers refactoring to reduce line length, if required, over breaking lines solely for readability.

File: server/authorization.go
Comment (emschwar) [inline server/authorization.go:307]: Codecov indicates these lines are not covered by a test. Can you add a test, please?

File: server/authorization.go
Comment (emschwar) [inline server/authorization.go:623]: This branch doesn't seem to have any tests that cover it. Can you please add one?

**Review by emschwar — COMMENTED** [review]
Please add a few tests, and I think we can omit breaking lines for readability purposes here. Generally, if we must have such long lines, the solution is to refactor them in order to not be so long, rather than to break the lines up solely for readability purposes.

**Review by theganyo — APPROVED** [review]

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go]: Done.

File: server/authorization.go
Comment (kansalarchit) [reply server/authorization.go:529]: This is currently not reachable, its just a safety check we have put in place.

**Comment by kansalarchit** [comment]
As per offline discussion with Eric, Removing the line-breaking call-site documentation and going back to refactored use-case specific methods for `unavailable()` and `internalError()`.
Leaving the rest as it is because it doesn't make any difference for readability.
Also added the unit-test as suggested.

Please take a look. Thanks!

File: server/header_context.go
Comment (emschwar) [inline server/header_context.go:109]: Hrm, you know you can define a method on `AdapterFault` here that can respond to `nil`? Something like:

```
func (af *AdapterFault) GetRpcCode() rpc.Code {
  if af == nil {
    return rpc.OK
  }

  return af.RpcCode
}
```

(Not saying you have to, but you can simplify these methods a bit if you like)

File: server/header_context.go
Comment (emschwar) [inline server/header_context.go:116]: Like here, you could define an accessor on `AdapterFault` that returns `"fault"` (or "defaultFaultCode", or whatever you like) if there is no `FaultCode` defined, and the value of `FaultCode` otherwise

**Review by emschwar — APPROVED** [review]
LGTM. I saw some places you could simplify things a bit, if you feel the need.

File: server/header_context.go
Comment (rockspore) [reply server/header_context.go:109]: Just wanted to point out it's probably better to do `RpcCode()` since `GetXYZ` is not recommended. I know there are bad examples in this repo as well as the proto-generated pkgs though.

Comment (emschwar) [reply server/header_context.go:109]: Oh gosh, yes! Thanks, I forgot that `GetFoo()` is an antipattern in Go.

Comment (kansalarchit) [reply server/header_context.go:109]: Thanks for the suggestions!
I think method and field with same name cannot co-exist. Skipping this for now.

File: server/header_context.go
Comment (kansalarchit) [reply server/header_context.go:116]: Thanks for the suggestion, I am not sure if I'd want adapterFault to return `fault` as the default value. The only reason I am keeping this here is to keep the existing logic unchanged.
Skipping this for now.

**Pull request merged**
